Code Review June'19

jkreindl opened the pull request.

pichristoph committed:
Current version (including wrong NodeFactory etc.) of debugExpr Parser

pichristoph committed:
Deleted wrong structures, inserted basic arithmetic AST operations

pichristoph committed:
implemented shifts, relExpr and ternary (if?then:else) operator.
Expressions now work with 'int' types.

pichristoph committed:
created new branch 'feature/debugexpr'

pichristoph committed:
Implemented ShortCircuitEvaluation

pichristoph committed:
implemented part of UnaryExpressions, adjusted DebugExprException class,
removed obsolete class 'DebugExprSymbolTable'

pichristoph committed:
Implemented error message handling

pichristoph committed:
improved error handling

pichristoph committed:
started test classes (without 'real' check) using LLVMTestBase.java.
Will change after the next commit(s).

pichristoph committed:
added test files

pichristoph committed:
implemented Tests for evaluating debug expressions
(LLVMDebugExprParserTest)

pichristoph committed:
added new tests and expanded existing tests

pichristoph committed:
implemented (double) floating point for symbol table, added floating
point tests

pichristoph committed:
Implemented floating point parsing and floating point comparison

pichristoph committed:
implemented string representation of boolean expressions in AST root

pichristoph committed:
added a part of type information in parser, (too) simple type tests

pichristoph committed:
implemented sizeof-Node

pichristoph committed:
added build for test files

pichristoph committed:
added LL(1) resolver for type casts

pichristoph committed:
Implemented simple type casts, still problem with unsigned casts and
chars

pichristoph committed:
added sign information for floating point casts

pichristoph committed:
created NodeFactory, removed DebugExprRootNode

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/LLVMLanguage.java:
```diff
@@ -132,6 +140,40 @@ protected void initializeContext(LLVMContext context) {
+                try {
+                    // for boolean results: convert to a string
+                    boolean result = root.executeI1(frame);
+                    return result ? "true" : "false";
```
Why not use `String.valueOf(result)`?

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprCompareNode.java:
```diff
@@ -0,0 +1,160 @@
+        if (cop == null)
+            return Parser.errorObjNode.executeGeneric(frame);
+        /* null is passed as type, since a type check is done by the arithmetic node anyway */
+        LLVMExpressionNode ret = nodeFactory.createComparison(cop, null, left, right);
```
As a rule, nodes should never be created on the general execution path. Also, each node must have only one parent in the AST, and that parent should not change.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprCompareNode.java:
```diff
@@ -0,0 +1,160 @@
+        } catch (UnsupportedSpecializationException e) {
+
+        }
+        cop = getFloatingPointOp();
```
This block of code is exactly the same as the next one.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprCompareNode.java:
```diff
@@ -0,0 +1,160 @@
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+import static com.oracle.truffle.llvm.runtime.CompareOperator.*;
+
+public class DebugExprCompareNode extends LLVMExpressionNode {
```
You are using a pattern here where essentially you check which type your argument values have. This is exactly what `@Specialization` is for. It allows you to declare which implementation of this node should be executed for different types of input values.

I think there are two ways to implement this node:
1) Since Types of expressions should already be known in the parser (either from the accessed source symbol, the literal or the sub-expressions) you can just create the correct compare node in the parser and use it directly instead of wrapping it in this node class.
2) If you look at `BasicNodeFactory#createComparison`, it always returns an `LLVMAbstractCompareNode` for comparing scalar values (i.e. unless you pass a `VectorType` to the factory function). You can use `LLVMAbstractCompareNode#executeWithTarget` to execute the node with explicit values provided by you rather than a child node. Then you can just pass `null` instead of `left` or `right` to the constructor.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprCompareNode.java:
```diff
@@ -0,0 +1,160 @@
+        }
+    }
+
+    private CompareOperator getUnsignedIntOp() {
```
Wir haben ja schon besprochen, dass du ein spezielles Object für unsigned integers einführen könntest. Nach Überlegung macht das aber wenig Sinn. Deine Expression-Grammatik ist an C angelehnt und daher strongly typed. Das heißt, dass du im Parser schon jederzeit den Typ eines Symbols / eines Wertes wissen musst. Damit sollte die Unterscheidung, ob eine Operation als signed, unsigned oder floating point ausgeführt wird, schon im Parser klar sein. Konkret schaut dass dann so aus, dass jede Regel im Parser ein Tupel aus Typ des Ergebnis-Wertes und Node zur Erzeugung des Ergebnis-Wertes zurückgibt. Auf Basis dieses Tupels kann dann der Typ des umgebenden Ausdruckes bestimmt werden.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprSCENode.java:
```diff
@@ -0,0 +1,70 @@
+import com.oracle.truffle.llvm.runtime.debug.debugexpr.parser.Parser;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+
+public class DebugExprSCENode extends LLVMExpressionNode {
```
Class names should most of all be descriptive, even is that makes them longer. Please avoid using abbreviations in class names.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprSizeofNode.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    public Object executeGeneric(VirtualFrame frame) {
+        if (type != null)
+            return type.getBitSize() / 8;
```
We discussed about getting the correct byte-size via the context. You can make this class abstract and implement a method `@Specialization protected Object doGeneric(VirtualFrame frame, @CachedContext(LLVMLanguage.class) ContextReference<LLVMContext> ref) { ... }`. This way, the Truffle DSL will automatically generate a node for you that calls your single specialization with a context reference. In your specialization you can then use `LLVMContext#getSize(Type)` to get the correct size. More concretely (and for reasons that I won't get into here), please call `getSize` only in another method that is separate from your specialization and is annotated with `@TruffleBoundary`.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprTernaryNode.java:
```diff
@@ -0,0 +1,62 @@
+        boolean cond;
+        try {
+            cond = condition.executeI1(frame);
+        } catch (UnexpectedResultException e) {
```
`UnexpectedResultException` is not something we usually catch by hand. For a cleaner implementation you could annotate your class with an `@NodeChild(LLVMExpressionNode.class)`, make it abstract, and have a `@Specialization` method that takes a `boolean` parameter and executes the correct `@Child`. You can have in this class also one method that that takes a generic `Object` as parameter and is annotated with `@Fallback`. This method is invoked by the DSL if no specializations apply to the `@NodeChild`.

rschatz replied:
Yes, that would have been easier. It would look roughly like this:

```
@NodeChild
public abstract class DebugExprTernaryNode extends LLVMExpressionNode {
    @Child private LLVMExpressionNode thenNode;
    @Child private LLVMExpressionNode elseNode;

    ...

    @Specialization
    Object doTernary(VirtualFrame frame, boolean condition) {
        if (conditionProfile.profile(condition)) {
            return thenNode.executeGeneric(frame);
        ....
    }

    @Fallback
    Object doError(Object condition) {
        throw DebugExprException.typeError(this, condition);
    }
}
```

The DSL generated code would look pretty much exactly like you wrote. The DSL maybe has the advantage that it provides additional profiling, but since the fallback case is an error case anyway, that doesn't make that much of a difference. The code just looks nicer.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java:
```diff
@@ -0,0 +1,95 @@
+        LLVMLanguage.getLLVMContextReference().get();
+        for (Scope scope : scopes) {
+            Object vars = scope.getVariables();
+            InteropLibrary library = InteropLibrary.getFactory().getUncached();
```
You can move `library` out of the loop.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java:
```diff
@@ -0,0 +1,95 @@
+            try {
+                final Object memberKeys = library.getMembers(vars);
+                library.hasArrayElements(memberKeys);
+                for (long i = 0; i < library.getArraySize(memberKeys); i++) {
```
Instead of iterating over the existing members, you can just call `library.isMemberReadable(vars, name)` to check if a scope has a member with a given name. Then you also do not need to catch an `InvalidArrayIndexException`.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java:
```diff
@@ -0,0 +1,95 @@
+                    if (!memberKey.equals(name))
+                        continue;
+                    Object member = library.readMember(vars, memberKey);
+                    LLVMDebuggerValue ldv = (LLVMDebuggerValue) member;
```
The member could also not currently have a value (e.g. if the user entered a symbol that the compiler removed and is therefore printed as `<unavailable>` or similar). Please make sure to handle the `ClassCastException` that would occur in such a case.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java:
```diff
@@ -0,0 +1,95 @@
+                    }
+                    // System.out.println(name + "|metaObj.toString() = " + metaObj.toString());
+                    // System.out.println(name + "|ldv.toString() = " + ldv.toString());
+                    if (metaObj.toString().contentEquals("int")) {
```
Languages other than C may not use these strings as type names. The `metaObj` for primitive values from any language is a `LLVMSourceBasicType`, which exposes `getKind` and `getSize` methods that you can use to determine how to parse the value. Please know that `getSize` in this case is the bit-size of the type, and that in C you can easily have integer types that are not the usual power-of-two sizes. In that case you can use `if (type.getSize() < Long.SIZE) { Long.parseLong(str) } else if (type.getSize() < Integer.SIZE) { ...` to choose the correct representation (Widening the types in this case is fine with me).

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExpr.atg:
```diff
@@ -0,0 +1,396 @@
+private LLVMExpressionNode astRoot=null;
+private ContextReference<LLVMContext> contextReference=null;
+private DebugExprNodeFactory NF;
+public final static DebugExprErrorNode noObjNode = DebugExprErrorNode.create("<cannot find expression>");
```
Please have these as part of the `NF` instead.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExpr.atg:
```diff
@@ -0,0 +1,396 @@
+// Example:
+//   Empty = "BEGIN" { Statement } "END".
+//   Statement = ident "=" number { "+" number } .
+DebugExpr		(. LLVMExpressionNode n=null; .)
```
Please re-format this ATG so that all code attributes (the `(. ... .)`) begin at the same column.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExpr.atg:
```diff
@@ -0,0 +1,396 @@
+=
+Designator<out n>
+|
+UnaryOp<out kind> CastExpr<out n> 		(. switch(kind) {
```
Please move this switch also to the `NF`.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExpr.atg:
```diff
@@ -0,0 +1,396 @@
+=
+CastExpr<out n> 
+{
+	 "*" CastExpr<out n1> 				(./*null is passed as type, since a type check is done by the arithmetic node anyway*/
```
The comment that you use `null` instead of a type because the type check is not necessary should be put in the `NF` rather than this grammar.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExpr.atg:
```diff
@@ -0,0 +1,396 @@
+[
+	IF (IsCast()) "(" DType<out typeP> ")" 
+]
+UnaryExpr<out n>			(. if(typeP!=null) {
```
If it takes more than a single line of code, it should be in the `NF`.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExprException.java:
```diff
@@ -0,0 +1,46 @@
+
+public class DebugExprException extends RuntimeException {
+
+    /**
```
Please remove empty JavaDoc.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExprParser.java:
```diff
@@ -0,0 +1,74 @@
+        parser.SetContextReference(contextReference);
+    }
+
+    public LLVMExpressionNode parse() throws Exception {
```
I'm not a fan of throwing generic `Exception` when you have knowledge of which exceptions can actually occur.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.test/src/com/oracle/truffle/llvm/test/debug/LLVMDebugExprParserTest.java:
```diff
@@ -0,0 +1,236 @@
+package com.oracle.truffle.llvm.test.debug;
```
Much of this class is copied from `LLVMDebugTestBase`. Ideally, we could merge some of the duplicated code.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.test/src/com/oracle/truffle/llvm/test/debug/LLVMDebugExprParserTest.java:
```diff
@@ -0,0 +1,236 @@
+            for (Entry<String, String> kv : textExpressionMap.entrySet()) {
+                String actual = frame.eval(kv.getKey()).as(String.class);
+                System.out.println("\"" + kv.getValue() + "\" vs. \"" + actual + "\"");
+                assertEquals("Expressions are not equal", kv.getValue(), actual);
```
This error message should rather read something like `Evaluation of expression 'xyz' produced unexpected result`.

jkreindl commented on sulong/runBcFile.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
Your own personal scripts and temporary tests should not be committed to the repository. Tests you want to keep should be put in `tests/*`. Also, please never commit compilation output if it is at all avoidable.

jkreindl reviewed (commented):
I was intentionally rather strict when reviewing this. Overall the code looks alright though.

I noticed that you haven't used the DSL at all. However, in some places it would make your code much cleaner. I have added comments where I think this is the case, and put some suggestions on how to fix. If you have any questions to this, please don't hesitate to ask.

pichristoph committed:
Implemented Type Node and common Expression node

pichristoph committed:
based on Code Review June 2019 #1: started applying changes concerning:
Move actions from Parser/ATG to NodeFactory
Restructured DebugExprVarNode s.t. a type of a variable is already known
while building the AST

pichristoph committed:
Implemented DSL for SCE, !, ~ operator

pichristoph committed:
added type information to grammar/RD parser methods
added methods for DebugExprType

pichristoph committed:
reorganized compare nodes

pichristoph committed:
applied DSL in TernaryNode, SizeofNode. Reformatted ATG. Sizeof(...) is
currently not working. Modified test messages.

pichristoph committed:
removed singleton pattern from class DebugExprNodeFactory

fixed bug as @field Iterable<Scope> scopes is now updated regularly

pichristoph committed:
Simplified evaluation in ExecutableNode, added possibility to get object
members (incorrect) and function calls (also incorrect)

pichristoph committed:
enabled function calls and accessing object members

also adjusted @NodeChild annotations with type information

pichristoph committed:
implemented basic array access

pichristoph committed:
enabled inner array evaluation (e.g. a[a[2]])

pichristoph committed:
Removed error node, implemented Exception catch.

Applied Short circuit evaluation and @Fallback (if no specialization
works)

pichristoph committed:
adjusted member access (object.member)

pichristoph committed:
removed @Children annotation from field List<Node> in FunctionCallNode

pichristoph committed:
adjusted tests for invalid expressions, disabled console output for
working tests

pichristoph committed:
updated tests

pichristoph committed:
removed unnecessary test files

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprFunctionCallNode.java:
```diff
@@ -0,0 +1,106 @@
+    private final String functionName;
+    private DebugExprType type;
+    private Object member;
+    private List<LLVMExpressionNode> arguments;
```
Child nodes should in general be contained in an `final` array that is annotated with `@Children`. Truffle partial evaluation cannot easily inline them otherwise.

jkreindl commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprFunctionCallNode.java:
```diff
@@ -0,0 +1,106 @@
+                member = library.readMember(vars, functionName);
+                if (library.isExecutable(member)) {
+                    try {
+                        Object[] argumentArr = arguments.stream().map(n -> n.executeGeneric(frame)).toArray();
```
Please avoid streams in `execute*` methods and in `@Specialization` methods. Streams allow for small and concise code, but are terrible for partial evaluation.

pichristoph committed:
Extended error messages for debugger, added @Children to
FunctionCallNode for AST

pichristoph committed:
added simple array tests

zapster commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprBitFlipNode.java:
```diff
@@ -0,0 +1,53 @@
+package com.oracle.truffle.llvm.runtime.debug.debugexpr.nodes;
```
Copyright header missing. `mx` might be able to help you: try `mx checkcopyrights --primary -- --fix` in the `sulong` directory. (but be sure to save/commit all changes before ;))

zapster commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprFunctionCallNode.java:
```diff
@@ -0,0 +1,109 @@
+                        return library.execute(member, argumentArr);
+                    } catch (UnsupportedTypeException e) {
+                        // TODO Auto-generated catch block
+                        e.printStackTrace();
```
I'd also just rethrow, or maybe not even catch at all.

rschatz replied:
Yes, definitely throw an exception here (rethrow doesn't work because it's a checked exception, but I guess we should just throw a `DebugExprException` here).

In general (here and also in a few other places, I'm not going to comment on all of them):
`printStackTrace()` is fine for local debugging, but please don't output anything to the console in the final code.

zapster commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprNodeFactory.java:
```diff
@@ -0,0 +1,309 @@
+import com.oracle.truffle.llvm.runtime.types.PrimitiveType;
+
+public final class DebugExprNodeFactory {
+    private ContextReference<LLVMContext> contextReference;
```
not sure you need to/should store the context reference here. You can always get it from `LLVMLanguage#getLLVMContextReference`. @rschatz?

rschatz replied:
I think it's fine. As far as I can tell, this object only lives for a single parse, so you can see this field as a kind of "cache" for efficiency. `LLVMLanguge.getLLVMContextReference` always has to look up a thread local field.

I would even go so far and store the `LLVMContext` here directly. In AST nodes that's not allowed, but here it would be fine, since this object is not persistent anyway.

zapster commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprNotNode.java:
```diff
@@ -0,0 +1,52 @@
+    public abstract Object executeWithTarget(Object child);
+
+    @NodeInfo(shortName = "!")
+    public abstract static class NotNode extends DebugExprBitFlipNode {
```
I think this should extend `DebugExprNotNode`. Btw, why are you using static inner classes instead of adding the `@Specialization`s to the `DebugExprNotNode` directly? Am I missing something?

rschatz replied:
Yes, I guess this could just be a single node, i.e. just put the `@NodeInfo` on the `DebugExprNotNode` and remove the inner class.

zapster commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprShortCircuitEvaluationNode.java:
```diff
@@ -0,0 +1,84 @@
+
+    /**
+     * Calculates the result of the short circuit operation. If the right node is not evaluated then
+     * <code>false</code> is provided.
```
minor: ~provided~ -> _returned_

rschatz replied:
That's not what is meant, right? What the comment tries to say is if the short-circuit rules say that the right value doesn't need to be evaluated (by returning `isEvaluateRight(leftValue) == false`), then this method will be called with `rightValue == false` (and by definition of the short-circuit rules, it shouldn't actually matter what value it's called with).

Maybe reword that comment a bit, e.g. explicitly stating that "if `isEvaluateRight` returns false, this will be called with `rightValue == false`". It took me some time to figure that out ;)

zapster commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprSizeofNode.java:
```diff
@@ -0,0 +1,60 @@
+
+public abstract class DebugExprSizeofNode extends LLVMExpressionNode {
+
+    private DebugExprType type;
```
could be `final`

zapster commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java:
```diff
@@ -0,0 +1,137 @@
+    @Override
+    public Object executeGeneric(VirtualFrame frame) {
+        findValue();
+        if (value == null)
```
We usually always use curly braces `{}`, even for single line branches. If you are very brave, you can try executing `mx checkstyle` and `mx findbugs`, but don't feel obligated to do so. ;)

jkreindl replied:
Actually, please do feel obligated to run these at least before you consider your project done. These commands must not find an error in order for your code to be able to pass the checks in the Travis and Oracle-internal gates. Although you want to run them with the `--primary` argument. This limits them to Sulong so they take less time and do not complain about bugs in other projects such as Truffle, which are not your concern anyways. Regarding the braces, I think `mx eclipseformat --primary` may do that for you.

zapster reviewed (commented):
Looks very good to me. Added a few comments, but all just minor remarks/hints. I'm especially impressed by, and happy about, the tests. :)

Note that I needed a small fix to get `mx gate --tags debugexpr` running locally (see zapster@5f7701f2ae6d23aabbb294cad726defbcb4c4a96, feel free to cherry-pick/copy-paste).

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprShortCircuitEvaluationNode.java:
```diff
@@ -0,0 +1,84 @@
+        return execute(leftValue, rightValue);
+    }
+
+    protected abstract boolean isEvaluateRight(boolean leftValue);
```
Minor: Better `shouldEvaluateRight`?

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprNotNode.java:
```diff
@@ -0,0 +1,52 @@
+
+@NodeChild(value = "child", type = LLVMExpressionNode.class)
+public abstract class DebugExprNotNode extends LLVMExpressionNode {
+    public abstract Object executeWithTarget(Object child);
```
Looks unused.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprBitFlipNode.java:
```diff
@@ -0,0 +1,53 @@
+
+@NodeChild(value = "child", type = LLVMExpressionNode.class)
+public abstract class DebugExprBitFlipNode extends LLVMExpressionNode {
+    public abstract Object executeWithTarget(Object child);
```
Looks unused.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprBitFlipNode.java:
```diff
@@ -0,0 +1,53 @@
+    public abstract Object executeWithTarget(Object child);
+
+    @NodeInfo(shortName = "~")
+    public abstract static class BitFlipNode extends DebugExprBitFlipNode {
```
See comment on `DebugExprNotNode`, I think this inner class is unnecessary.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/LLVMLanguage.java:
```diff
@@ -132,6 +137,40 @@ protected void initializeContext(LLVMContext context) {
+
+                    try {
+                        // try to execute node
+                        return String.valueOf(root.executeGeneric(frame));
```
Truffle won't like you accessing the `root` variable from outside as it is. The problem is that the partial evaluator will not realize this is a Truffle node, because the variable is not annotated with `@Child`.

The solution is to make an explicit field in the inner class `@Child LLVMExpressionNode rootExpr = root`, and use `rootExpr` here instead of `root` directly.

Unfortunately there is no way to catch this error automatically, and it's probably still going to work fine in most of the cases. It probably breaks only in rather subtle ways, like making problems with graph dumps or with code instrumentation.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprSizeofNode.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @TruffleBoundary
+    protected int getSizeInformation(ContextReference<LLVMContext> ref) {
```
The `@TruffleBoundary` here is placed wrong. Since `type` is final (see comment from @zapster), `type.getLLVMRuntimeType()` should constant fold. But since it's behind a `@TruffleBoundary`, the compiler can't see it.

But actually, since `type` can't change, will this node not always return a constant? Can't you just run this code in the constructor, and just return the constant value at runtime?

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprSizeofNode.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Specialization
+    protected Object doGeneric(VirtualFrame frame, @CachedContext(LLVMLanguage.class) ContextReference<LLVMContext> ref) {
```
You can just do `@CachedContext(LLVMLanguage.class) LLVMContext ctx`. It's a shortcut, if you will call `get()` on the reference anyway.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/CocoInputStream.java:
```diff
@@ -0,0 +1,64 @@
+    private final String s;
+    private int pos;
+
+    public CocoInputStream(String s) {
```
Looks unused.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/CocoInputStream.java:
```diff
@@ -0,0 +1,64 @@
+ * Parser.
+ */
+public class CocoInputStream extends InputStream {
+    private final String s;
```
Why not make this a `CharSequence`? Then you don't need the `toString()` below.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExprException.java:
```diff
@@ -0,0 +1,94 @@
+        this.message = message;
+    }
+
+    public static final DebugExprException typeError(LLVMExpressionNode operation, Object... members) {
```
This should be a `@TruffleBoundary`. You're using this in various places in Truffle `execute` methods, and the StringBuilder stuff in here should probably not be inlined.

Same thing for the other factory methods.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExprParser.java:
```diff
@@ -0,0 +1,77 @@
+
+    public LLVMExpressionNode parse() throws DebugExprException {
+        final StringBuilder sb = new StringBuilder();
+        OutputStream errorStream = new OutputStream() {
```
Minor: It's probably more efficient to use a `ByteArrayOutputStream`, and convert the result to a String at the end.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExprType.java:
```diff
@@ -0,0 +1,294 @@
+
+    private final Kind kind;
+    private DebugExprType innerType; // used for arrays, pointers, ...
+    private static HashMap<Kind, DebugExprType> map = new HashMap<>();
```
When the key is an enum, you should use an `EnumMap` instead of a `HashMap`. It's a lot more efficient.

Or even better, since it's your own enum type, you can store the `DebugExprType` singleton directly in the enum, instead of using a map.. See `LLVMInteropType.ValueKind`, that's already using the same pattern.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExprType.java:
```diff
@@ -0,0 +1,294 @@
+        else if (t1.isFloatingType() && t2.isFloatingType()) {
+            return getFloatType(Math.max(t1.getBitSize(), t2.getBitSize()));
+        } else if (t1.isIntegerType() && t2.isIntegerType()) {
+            if (t1.isUnsigned() ^ t2.isUnsigned()) {
```
I terms of readability, I prefer `!=` ;)

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExprType.java:
```diff
@@ -0,0 +1,294 @@
+        }
+    }
+
+    public static DebugExprType commonType(DebugExprType t1, DebugExprType t2) {
```
This method could use a comment what it's supposed to do. I guess it's returning the more general type in terms of implicit conversions?

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprNodeFactory.java:
```diff
@@ -0,0 +1,309 @@
+    public DebugExpressionPair createArithmeticOp(ArithmeticOperation op, DebugExpressionPair left, DebugExpressionPair right) {
+        checkError(left, op.name());
+        checkError(right, op.name());
+        /* null is passed as type, since a type check is done by the arithmetic node anyway */
```
That is not true anymore, on current master you have to pass a type. But I guess in that case you can just pass `commonType.getLLVMRuntimeType()`, right?

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java:
```diff
@@ -0,0 +1,137 @@
+        InteropLibrary library = InteropLibrary.getFactory().getUncached();
+        for (Scope scope : scopes) {
+            Object vars = scope.getVariables();
+            library.hasMembers(vars);
```
Looks unnecessary. As it is, this line does nothing, since its return value is not used. And `hasMembers` should have no side-effect (if it has, it's a bug we should fix in Sulong ;) ).

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java:
```diff
@@ -0,0 +1,137 @@
+                    type = DebugExprType.getTypeFromSymbolTableMetaObject(metaObj);
+                    return;
+                }
+                if (library.hasArrayElements(memberKeys) && library.isMemberReadable(memberKeys, name)) {
```
That looks wrong. `memberKeys` should always be an array (but no harm in checking), but then `isMemberReadable` will probably always return `false`, since `memberKeys` is an array, not a struct. Are you sure this `if` is ever entered?

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java:
```diff
@@ -0,0 +1,137 @@
+
+    private final String name;
+    private Iterable<Scope> scopes;
+    private DebugExprType type;
```
Should this really be here? Shouldn't this rather be in the `DebugExpressionPair`? That way you also wouldn't need the type distinctions in the `DebugExprNodeFactory` (see my comment there).

EDIT: I realized that you already store it in `DebugExpressionPair`, too.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprNodeFactory.java:
```diff
@@ -0,0 +1,309 @@
+        DebugExprType baseType = null;
+        if (array.getNode() instanceof DebugExprVarNode) {
+            baseMember = ((DebugExprVarNode) (array.getNode())).getMember();
+            baseType = ((DebugExprVarNode) (array.getNode())).getType();
```
Why is this not just using `array.getType()`? It should always be the same, and that way, you don't need the `if-instanceof` cascade at all (see comment below why I think you don't need the `baseMember`).

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprNodeFactory.java:
```diff
@@ -0,0 +1,309 @@
+            baseType = ((DebugExprObjectMemberNode) (array.getNode())).getType();
+        }
+        if (baseMember != null) {
+            DebugExprArrayElementNode node = new DebugExprArrayElementNode(baseMember, index.getNode(), baseType.getInnerType());
```
I don't think the `DebugExprArrayElementNode` should get the `baseMember` as argument, it should rather have the `array.getNode()` as a child node. You shouldn't store runtime values in the AST. It's not really avoidable in the `DebugExprVarNode`, but we should at least avoid it where possible. In this case, the `DebugExprArrayElementNode` can call `execute` on the `array.getNode()` at runtime.

See also the comment in `DebugExprArrayElementNode`.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprNodeFactory.java:
```diff
@@ -0,0 +1,309 @@
+    }
+
+    @SuppressWarnings("static-method")
+    public DebugExpressionPair createObjectMember(DebugExpressionPair receiver, String fieldName) {
```
See comments in `createArrayElement`, exactly the same applies here.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprArrayElementNode.java:
```diff
@@ -0,0 +1,101 @@
+    @Child private LLVMExpressionNode indexNode;
+    private DebugExprType type;
+    private Object member;
+    private Object baseMember;
```
I don't think these should be stored in the runtime node. Rather, the node should have another child node `@Child private LLVMExpressionNode arrayNode`. Then, `executeGeneric` would just do `arrayNode.executeGeneric()` to get the array.

The same applies to the `DebugExprObjectMemberNode`.

Btw. `member` is not even written anywhere, and neither `getType()` nor `getMember()` are called anywhere.

rschatz commented on sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprArrayElementNode.java:
```diff
@@ -0,0 +1,101 @@
+        if (library.hasMembers(baseMember)) {
+            Object getmembers;
+            try {
+                getmembers = library.getMembers(baseMember);
```
@jkreindl is this really correct? Do our debug arrays really expose their array elements as members? Why not as array elements?

jkreindl replied:
I don't remember about the reason for that. Maybe the debugger didn't use to show array elements at some point (I was probably still using the NetBeans plugin then), or maybe I decided that I wanted a particular formatting for the array indices that matched lldb. We could change that though.

rschatz replied:
Ok. No need to change it now, I was just wondering, because that code looks weird.

rschatz reviewed (commented):
Looks very good to me. I left a few comments to the code.

The only major thing are the comments around the array index and member name operations. Storing runtime values in an AST is always bad. The problem is that then you can't re-execute that code at a different point in time. I know the debugger currently doesn't do this currently, and always re-parses, but what if they suddenly decide to optimize this, and don't re-parse if the same watch expression is evaluated at the same program location?

I realize it's not so easy to resolve, especially interacting with our existing codebase, that's also not so nice in some places. I guess the main problem here is that you can't determine the static type of variables without actually reading them. I guess the "correct" way to resolve that is to read in the constructor or factory to determine the type, but then read again in the execute method (the type shouldn't actually change). That may sound inefficient, but it's usually not so bad in practice.

All other comments are just minor efficiency or style issues, and some Truffle technicalities that you really have no chance knowing.

In general, I'm impressed with the quality of the code. This looks like we can actually merge it with relatively little cleanup and fixes necessary.

pichristoph committed:
Fixed test path for debugexpr tests

pichristoph committed:
short circuit evaluation: renamed method, refactored comment

pichristoph committed:
applied DSL for DebugExprTernaryNode

pichristoph committed:
Removed inner classes from DebugExprBitFlipNode and DebugExprNotNode as
they are not needed

pichristoph committed:
added Copyright header to DebugExprBitFlipNode and DebugExprNotNode

pichristoph committed:
marked rootExpr as Truffle Node s.t. Truffle is able to optimize

pichristoph committed:
refactored "member" handling. Object member access (struct.member) is
currently not possible.

pichristoph committed:
expanded tests, removed unnecessary console output

pichristoph committed:
removed string option

pichristoph commented:
Thank you for reviewing my code (@jkreindl @zapster @rschatz)! It has/will become simpler and clearer because of your hints :)
For reasons of simplicity I do not/will not answer on each of your comments.

pichristoph committed:
enabled object member (struct) access, removed reference types (&)

pichristoph committed:
added character handling, (incorrect) DereferenceNode for pointer access

pichristoph committed:
extended DebugExprType class and function call types

pichristoph committed:
extended type handling, fixed minor errors

pichristoph committed:
added pointer test cases

pichristoph committed:
fix small errors of type handling, extend type class

pichristoph committed:
Add dereferencing (not working, 'LLVMDebugObject::getValue' is private)

pichristoph committed:
add pointer dereferencing (working)

pichristoph committed:
enable debugger inspection for root expressions of a non-primitive type

pichristoph committed:
removed test case which fails due to a clang error

pichristoph committed:
refactor 'typeof' and add pointer casts

pichristoph committed:
add new tests of 'testPointerStruct'

the old tests do not work due to an error in clang

pichristoph committed:
style fixes to pass checkStyle

pichristoph committed:
added style check information for generated classes

pichristoph committed:
add changes for rebase

pichristoph committed:
apply changes for rebase

pichristoph committed:
Add information of expression evaluation (draft)

A first draft describing the features and environmental settings for using Sulong's expression evaluation (watch expressions or command line).
